diff --git a/source/TR-181/sbapi/cosa_wifi_apis.c b/source/TR-181/sbapi/cosa_wifi_apis.c
index 1f90fab..2e03c7e 100644
--- a/source/TR-181/sbapi/cosa_wifi_apis.c
+++ b/source/TR-181/sbapi/cosa_wifi_apis.c
@@ -8964,7 +8964,7 @@ printf("%s: Reset FactoryReset to 0 \n",__FUNCTION__);
     CosaDmlWiFi_GetRapidReconnectIndicationEnable(&(pMyObject->bRapidReconnectIndicationEnabled), true);
     CosaDmlWiFiGetvAPStatsFeatureEnable(&(pMyObject->bX_RDKCENTRAL_COM_vAPStatsEnable));
     CosaDmlWiFiGetTxOverflowSelfheal(&(pMyObject->bTxOverflowSelfheal));
-#if !defined(_PLATFORM_RASPBERRYPI_) && !defined(_PLATFORM_TURRIS_)
+#if !defined(_PLATFORM_RASPBERRYPI_) && !defined(_PLATFORM_TURRIS_) && !defined(_COSA_BCM_ARM_)
     CosaDmlWiFiGetForceDisableWiFiRadio(&(pMyObject->bForceDisableWiFiRadio));
 #endif
 #if defined(FEATURE_HOSTAP_AUTHENTICATOR)
@@ -10189,7 +10189,7 @@ ANSC_STATUS CosaDmlWiFiGetForceDisableWiFiRadio(BOOLEAN *pbValue)
 
     // Initialize the value as FALSE always
     *pbValue = FALSE;
-#if defined(_PLATFORM_RASPBERRYPI_) || defined(_PLATFORM_TURRIS_)
+#if defined(_PLATFORM_RASPBERRYPI_) || defined(_PLATFORM_TURRIS_) || defined(_COSA_BCM_ARM_)
     UNREFERENCED_PARAMETER(strValue);
     return ANSC_STATUS_SUCCESS;
 #else
