diff --git a/release/src/router/openssl/demos/ssltest-ecc/ssltest.sh b/release/src/router/openssl/demos/ssltest-ecc/ssltest.sh
index 923ca43..2dc2273 100755
--- a/release/src/router/openssl/demos/ssltest-ecc/ssltest.sh
+++ b/release/src/router/openssl/demos/ssltest-ecc/ssltest.sh
@@ -36,7 +36,7 @@ ECDH_RSA_CIPHER_LIST="ECDH-RSA-AES256-SHA ECDH-RSA-AES128-SHA ECDH-RSA-DES-CBC3-
 
 # These require RSA certificates.
 # The RSA public key must be authorized for digital signature.
-ECDHE_RSA_CIPHER_LIST="ECDHE-RSA-AES256-SHA ECDHE-RSA-AES128-SHA ECDHE-RSA-DES-CBC3-SHA ECDHE-RSA-RC4-SHA ECDHE-RSA-NULL-SHA"
+ECDHE_RSA_CIPHER_LIST="ECDHE-RSA-AES256-SHA ECDHE-RSA-RC4-SHA ECDHE-RSA-NULL-SHA"
 
 # List of Elliptic curves over which we wish to test generation of
 # ephemeral ECDH keys when using AECDH or ECDHE ciphers
diff --git a/release/src/router/openssl/doc/apps/ciphers.pod b/release/src/router/openssl/doc/apps/ciphers.pod
index 9224557..9540c9d 100644
--- a/release/src/router/openssl/doc/apps/ciphers.pod
+++ b/release/src/router/openssl/doc/apps/ciphers.pod
@@ -372,7 +372,6 @@ e.g. DES-CBC3-SHA. In these cases, RSA authentication is used.
  SSL_RSA_WITH_IDEA_CBC_SHA               IDEA-CBC-SHA
  SSL_RSA_EXPORT_WITH_DES40_CBC_SHA       EXP-DES-CBC-SHA
  SSL_RSA_WITH_DES_CBC_SHA                DES-CBC-SHA
- SSL_RSA_WITH_3DES_EDE_CBC_SHA           DES-CBC3-SHA
 
  SSL_DH_DSS_WITH_DES_CBC_SHA             DH-DSS-DES-CBC-SHA
  SSL_DH_DSS_WITH_3DES_EDE_CBC_SHA        DH-DSS-DES-CBC3-SHA
@@ -429,8 +428,6 @@ e.g. DES-CBC3-SHA. In these cases, RSA authentication is used.
 
 =head2 AES ciphersuites from RFC3268, extending TLS v1.0
 
- TLS_RSA_WITH_AES_128_CBC_SHA            AES128-SHA
- TLS_RSA_WITH_AES_256_CBC_SHA            AES256-SHA
 
  TLS_DH_DSS_WITH_AES_128_CBC_SHA         DH-DSS-AES128-SHA
  TLS_DH_DSS_WITH_AES_256_CBC_SHA         DH-DSS-AES256-SHA
@@ -511,8 +508,6 @@ Note: these ciphers can also be used in SSL v3.
 
  TLS_ECDHE_RSA_WITH_NULL_SHA             ECDHE-RSA-NULL-SHA
  TLS_ECDHE_RSA_WITH_RC4_128_SHA          ECDHE-RSA-RC4-SHA
- TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA     ECDHE-RSA-DES-CBC3-SHA
- TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA      ECDHE-RSA-AES128-SHA
  TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA      ECDHE-RSA-AES256-SHA
 
  TLS_ECDHE_ECDSA_WITH_NULL_SHA           ECDHE-ECDSA-NULL-SHA
diff --git a/release/src/router/openssl/doc/ssleay.txt b/release/src/router/openssl/doc/ssleay.txt
index c9b29bd..b871051 100644
--- a/release/src/router/openssl/doc/ssleay.txt
+++ b/release/src/router/openssl/doc/ssleay.txt
@@ -6063,7 +6063,6 @@ RC4-SHA                 SSLv3 Kx=RSA      Au=RSA  Enc=RC4(128)  Mac=SHA1
 RC4-MD5                 SSLv3 Kx=RSA      Au=RSA  Enc=RC4(128)  Mac=MD5 
 EDH-RSA-DES-CBC3-SHA    SSLv3 Kx=DH       Au=RSA  Enc=3DES(168) Mac=SHA1
 EDH-DSS-DES-CBC3-SHA    SSLv3 Kx=DH       Au=DSS  Enc=3DES(168) Mac=SHA1
-DES-CBC3-SHA            SSLv3 Kx=RSA      Au=RSA  Enc=3DES(168) Mac=SHA1
 IDEA-CBC-MD5            SSLv3 Kx=RSA      Au=RSA  Enc=IDEA(128) Mac=SHA1
 EDH-RSA-DES-CBC-SHA     SSLv3 Kx=DH       Au=RSA  Enc=DES(56)   Mac=SHA1
 EDH-DSS-DES-CBC-SHA     SSLv3 Kx=DH       Au=DSS  Enc=DES(56)   Mac=SHA1
diff --git a/release/src/router/openssl/ssl/s3_lib.c b/release/src/router/openssl/ssl/s3_lib.c
index 0385e03..0e9dfd9 100644
--- a/release/src/router/openssl/ssl/s3_lib.c
+++ b/release/src/router/openssl/ssl/s3_lib.c
@@ -320,7 +320,7 @@ OPENSSL_GLOBAL SSL_CIPHER ssl3_ciphers[] = {
 #endif
 
 /* Cipher 0A */
-    {
+    /*{
      1,
      SSL3_TXT_RSA_DES_192_CBC3_SHA,
      SSL3_CK_RSA_DES_192_CBC3_SHA,
@@ -333,7 +333,7 @@ OPENSSL_GLOBAL SSL_CIPHER ssl3_ciphers[] = {
      SSL_HANDSHAKE_MAC_DEFAULT | TLS1_PRF,
      112,
      168,
-     },
+     },*/
 
 /* The DH ciphers */
 /* Cipher 0B */
@@ -927,7 +927,7 @@ OPENSSL_GLOBAL SSL_CIPHER ssl3_ciphers[] = {
 
 /* New AES ciphersuites */
 /* Cipher 2F */
-    {
+    /*{
      1,
      TLS1_TXT_RSA_WITH_AES_128_SHA,
      TLS1_CK_RSA_WITH_AES_128_SHA,
@@ -940,7 +940,7 @@ OPENSSL_GLOBAL SSL_CIPHER ssl3_ciphers[] = {
      SSL_HANDSHAKE_MAC_DEFAULT | TLS1_PRF,
      128,
      128,
-     },
+     },*/
 /* Cipher 30 */
     {
      1,
@@ -1018,7 +1018,7 @@ OPENSSL_GLOBAL SSL_CIPHER ssl3_ciphers[] = {
      },
 
 /* Cipher 35 */
-    {
+    /*{
      1,
      TLS1_TXT_RSA_WITH_AES_256_SHA,
      TLS1_CK_RSA_WITH_AES_256_SHA,
@@ -1031,7 +1031,7 @@ OPENSSL_GLOBAL SSL_CIPHER ssl3_ciphers[] = {
      SSL_HANDSHAKE_MAC_DEFAULT | TLS1_PRF,
      256,
      256,
-     },
+     },*/
 /* Cipher 36 */
     {
      1,
@@ -2351,7 +2351,7 @@ OPENSSL_GLOBAL SSL_CIPHER ssl3_ciphers[] = {
      },
 
     /* Cipher C012 */
-    {
+    /*{
      1,
      TLS1_TXT_ECDHE_RSA_WITH_DES_192_CBC3_SHA,
      TLS1_CK_ECDHE_RSA_WITH_DES_192_CBC3_SHA,
@@ -2364,10 +2364,10 @@ OPENSSL_GLOBAL SSL_CIPHER ssl3_ciphers[] = {
      SSL_HANDSHAKE_MAC_DEFAULT | TLS1_PRF,
      112,
      168,
-     },
+     },*/
 
     /* Cipher C013 */
-    {
+    /*{
      1,
      TLS1_TXT_ECDHE_RSA_WITH_AES_128_CBC_SHA,
      TLS1_CK_ECDHE_RSA_WITH_AES_128_CBC_SHA,
@@ -2380,7 +2380,7 @@ OPENSSL_GLOBAL SSL_CIPHER ssl3_ciphers[] = {
      SSL_HANDSHAKE_MAC_DEFAULT | TLS1_PRF,
      128,
      128,
-     },
+     },*/
 
     /* Cipher C014 */
     {
diff --git a/release/src/router/openssl/ssl/ssl.h b/release/src/router/openssl/ssl/ssl.h
index 90aeb0c..3511442 100644
--- a/release/src/router/openssl/ssl/ssl.h
+++ b/release/src/router/openssl/ssl/ssl.h
@@ -187,7 +187,7 @@ extern "C" {
 # define SSL_TXT_DES_64_CBC_WITH_MD5     SSL2_TXT_DES_64_CBC_WITH_MD5
 # define SSL_TXT_DES_64_CBC_WITH_SHA     SSL2_TXT_DES_64_CBC_WITH_SHA
 # define SSL_TXT_DES_192_EDE3_CBC_WITH_MD5 SSL2_TXT_DES_192_EDE3_CBC_WITH_MD5
-# define SSL_TXT_DES_192_EDE3_CBC_WITH_SHA SSL2_TXT_DES_192_EDE3_CBC_WITH_SHA
+//# define SSL_TXT_DES_192_EDE3_CBC_WITH_SHA SSL2_TXT_DES_192_EDE3_CBC_WITH_SHA
 
 /*
  * VRS Additional Kerberos5 entries
diff --git a/release/src/router/openssl/ssl/ssl2.h b/release/src/router/openssl/ssl/ssl2.h
index 03c7dd8..0b24495 100644
--- a/release/src/router/openssl/ssl/ssl2.h
+++ b/release/src/router/openssl/ssl/ssl2.h
@@ -114,7 +114,6 @@ extern "C" {
 # define SSL2_TXT_DES_64_CBC_WITH_MD5            "DES-CBC-MD5"
 # define SSL2_TXT_DES_64_CBC_WITH_SHA            "DES-CBC-SHA"
 # define SSL2_TXT_DES_192_EDE3_CBC_WITH_MD5      "DES-CBC3-MD5"
-# define SSL2_TXT_DES_192_EDE3_CBC_WITH_SHA      "DES-CBC3-SHA"
 # define SSL2_TXT_RC4_64_WITH_MD5                "RC4-64-MD5"
 
 # define SSL2_TXT_NULL                           "NULL"
diff --git a/release/src/router/openssl/ssl/ssl3.h b/release/src/router/openssl/ssl/ssl3.h
index e681d50..741dee3 100644
--- a/release/src/router/openssl/ssl/ssl3.h
+++ b/release/src/router/openssl/ssl/ssl3.h
@@ -217,7 +217,6 @@ extern "C" {
 # define SSL3_TXT_RSA_IDEA_128_SHA               "IDEA-CBC-SHA"
 # define SSL3_TXT_RSA_DES_40_CBC_SHA             "EXP-DES-CBC-SHA"
 # define SSL3_TXT_RSA_DES_64_CBC_SHA             "DES-CBC-SHA"
-# define SSL3_TXT_RSA_DES_192_CBC3_SHA           "DES-CBC3-SHA"
 
 # define SSL3_TXT_DH_DSS_DES_40_CBC_SHA          "EXP-DH-DSS-DES-CBC-SHA"
 # define SSL3_TXT_DH_DSS_DES_64_CBC_SHA          "DH-DSS-DES-CBC-SHA"
diff --git a/release/src/router/openssl/ssl/t1_trce.c b/release/src/router/openssl/ssl/t1_trce.c
index c5e21df..c5ddea0 100644
--- a/release/src/router/openssl/ssl/t1_trce.c
+++ b/release/src/router/openssl/ssl/t1_trce.c
@@ -152,7 +152,7 @@ static ssl_trace_tbl ssl_ciphers_tbl[] = {
     {0x0007, "SSL_RSA_WITH_IDEA_CBC_SHA"},
     {0x0008, "SSL_RSA_EXPORT_WITH_DES40_CBC_SHA"},
     {0x0009, "SSL_RSA_WITH_DES_CBC_SHA"},
-    {0x000A, "SSL_RSA_WITH_3DES_EDE_CBC_SHA"},
+    //{0x000A, "SSL_RSA_WITH_3DES_EDE_CBC_SHA"},
     {0x000B, "SSL_DH_DSS_EXPORT_WITH_DES40_CBC_SHA"},
     {0x000C, "SSL_DH_DSS_WITH_DES_CBC_SHA"},
     {0x000D, "SSL_DH_DSS_WITH_3DES_EDE_CBC_SHA"},
@@ -185,13 +185,13 @@ static ssl_trace_tbl ssl_ciphers_tbl[] = {
     {0x0029, "TLS_KRB5_EXPORT_WITH_DES_CBC_40_MD5"},
     {0x002A, "TLS_KRB5_EXPORT_WITH_RC2_CBC_40_MD5"},
     {0x002B, "TLS_KRB5_EXPORT_WITH_RC4_40_MD5"},
-    {0x002F, "TLS_RSA_WITH_AES_128_CBC_SHA"},
+    //{0x002F, "TLS_RSA_WITH_AES_128_CBC_SHA"},
     {0x0030, "TLS_DH_DSS_WITH_AES_128_CBC_SHA"},
     {0x0031, "TLS_DH_RSA_WITH_AES_128_CBC_SHA"},
     {0x0032, "TLS_DHE_DSS_WITH_AES_128_CBC_SHA"},
     {0x0033, "TLS_DHE_RSA_WITH_AES_128_CBC_SHA"},
     {0x0034, "TLS_DH_anon_WITH_AES_128_CBC_SHA"},
-    {0x0035, "TLS_RSA_WITH_AES_256_CBC_SHA"},
+    //{0x0035, "TLS_RSA_WITH_AES_256_CBC_SHA"},
     {0x0036, "TLS_DH_DSS_WITH_AES_256_CBC_SHA"},
     {0x0037, "TLS_DH_RSA_WITH_AES_256_CBC_SHA"},
     {0x0038, "TLS_DHE_DSS_WITH_AES_256_CBC_SHA"},
@@ -300,8 +300,8 @@ static ssl_trace_tbl ssl_ciphers_tbl[] = {
     {0xC00F, "TLS_ECDH_RSA_WITH_AES_256_CBC_SHA"},
     {0xC010, "TLS_ECDHE_RSA_WITH_NULL_SHA"},
     {0xC011, "TLS_ECDHE_RSA_WITH_RC4_128_SHA"},
-    {0xC012, "TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA"},
-    {0xC013, "TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA"},
+    //{0xC012, "TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA"},
+    //{0xC013, "TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA"},
     {0xC014, "TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA"},
     {0xC015, "TLS_ECDH_anon_WITH_NULL_SHA"},
     {0xC016, "TLS_ECDH_anon_WITH_RC4_128_SHA"},
diff --git a/release/src/router/openssl/ssl/tls1.h b/release/src/router/openssl/ssl/tls1.h
index 7e237d0..5a5c402 100644
--- a/release/src/router/openssl/ssl/tls1.h
+++ b/release/src/router/openssl/ssl/tls1.h
@@ -578,14 +578,14 @@ SSL_CTX_callback_ctrl(ssl,SSL_CTRL_SET_TLSEXT_TICKET_KEY_CB,(void (*)(void))cb)
 # define TLS1_TXT_DHE_DSS_WITH_RC4_128_SHA               "DHE-DSS-RC4-SHA"
 
 /* AES ciphersuites from RFC3268 */
-# define TLS1_TXT_RSA_WITH_AES_128_SHA                   "AES128-SHA"
+//# define TLS1_TXT_RSA_WITH_AES_128_SHA                   "AES128-SHA"
 # define TLS1_TXT_DH_DSS_WITH_AES_128_SHA                "DH-DSS-AES128-SHA"
 # define TLS1_TXT_DH_RSA_WITH_AES_128_SHA                "DH-RSA-AES128-SHA"
 # define TLS1_TXT_DHE_DSS_WITH_AES_128_SHA               "DHE-DSS-AES128-SHA"
 # define TLS1_TXT_DHE_RSA_WITH_AES_128_SHA               "DHE-RSA-AES128-SHA"
 # define TLS1_TXT_ADH_WITH_AES_128_SHA                   "ADH-AES128-SHA"
 
-# define TLS1_TXT_RSA_WITH_AES_256_SHA                   "AES256-SHA"
+//# define TLS1_TXT_RSA_WITH_AES_256_SHA                   "AES256-SHA"
 # define TLS1_TXT_DH_DSS_WITH_AES_256_SHA                "DH-DSS-AES256-SHA"
 # define TLS1_TXT_DH_RSA_WITH_AES_256_SHA                "DH-RSA-AES256-SHA"
 # define TLS1_TXT_DHE_DSS_WITH_AES_256_SHA               "DHE-DSS-AES256-SHA"
@@ -613,8 +613,8 @@ SSL_CTX_callback_ctrl(ssl,SSL_CTRL_SET_TLSEXT_TICKET_KEY_CB,(void (*)(void))cb)
 
 # define TLS1_TXT_ECDHE_RSA_WITH_NULL_SHA                "ECDHE-RSA-NULL-SHA"
 # define TLS1_TXT_ECDHE_RSA_WITH_RC4_128_SHA             "ECDHE-RSA-RC4-SHA"
-# define TLS1_TXT_ECDHE_RSA_WITH_DES_192_CBC3_SHA        "ECDHE-RSA-DES-CBC3-SHA"
-# define TLS1_TXT_ECDHE_RSA_WITH_AES_128_CBC_SHA         "ECDHE-RSA-AES128-SHA"
+//# define TLS1_TXT_ECDHE_RSA_WITH_DES_192_CBC3_SHA        "ECDHE-RSA-DES-CBC3-SHA"
+//# define TLS1_TXT_ECDHE_RSA_WITH_AES_128_CBC_SHA         "ECDHE-RSA-AES128-SHA"
 # define TLS1_TXT_ECDHE_RSA_WITH_AES_256_CBC_SHA         "ECDHE-RSA-AES256-SHA"
 
 # define TLS1_TXT_ECDH_anon_WITH_NULL_SHA                "AECDH-NULL-SHA"
diff --git a/release/src/router/openssl/test/testfipsssl b/release/src/router/openssl/test/testfipsssl
index b8bbb25..876f09d 100644
--- a/release/src/router/openssl/test/testfipsssl
+++ b/release/src/router/openssl/test/testfipsssl
@@ -11,7 +11,7 @@ else
   cert="$2"
 fi
 
-ciphers="DHE-RSA-AES256-SHA:DHE-DSS-AES256-SHA:AES256-SHA:EDH-RSA-DES-CBC3-SHA:EDH-DSS-DES-CBC3-SHA:DES-CBC3-SHA:DHE-RSA-AES128-SHA:DHE-DSS-AES128-SHA:AES128-SHA:EXP1024-DHE-DSS-DES-CBC-SHA:EXP1024-DES-CBC-SHA:EDH-RSA-DES-CBC-SHA:EDH-DSS-DES-CBC-SHA:DES-CBC-SHA:EXP-EDH-RSA-DES-CBC-SHA:EXP-EDH-DSS-DES-CBC-SHA:EXP-DES-CBC-SHA"
+ciphers="DHE-RSA-AES256-SHA:DHE-DSS-AES256-SHA:EDH-RSA-DES-CBC3-SHA:EDH-DSS-DES-CBC3-SHA:DHE-RSA-AES128-SHA:DHE-DSS-AES128-SHA:EXP1024-DHE-DSS-DES-CBC-SHA:EXP1024-DES-CBC-SHA:EDH-RSA-DES-CBC-SHA:EDH-DSS-DES-CBC-SHA:DES-CBC-SHA:EXP-EDH-RSA-DES-CBC-SHA:EXP-EDH-DSS-DES-CBC-SHA:EXP-DES-CBC-SHA"
 
 ssltest="../util/shlib_wrap.sh ./ssltest -F -key $key -cert $cert -c_key $key -c_cert $cert -cipher $ciphers"
 
diff --git a/release/src/router/openssl/test/testssl b/release/src/router/openssl/test/testssl
index a6f9fa7..65bf0e2 100644
--- a/release/src/router/openssl/test/testssl
+++ b/release/src/router/openssl/test/testssl
@@ -288,7 +288,6 @@ fi
 # Multi-buffer tests
 
 if [ -z "$extra" -a `uname -m` = "x86_64" ]; then
-  $ssltest -cipher AES128-SHA    -bytes 8m	|| exit 1
   $ssltest -cipher AES128-SHA256 -bytes 8m	|| exit 1
 fi
 
diff --git a/release/src/router/openssl/test/times b/release/src/router/openssl/test/times
index 6b66eb3..d1b2b4a 100644
--- a/release/src/router/openssl/test/times
+++ b/release/src/router/openssl/test/times
@@ -28,14 +28,12 @@ SSLv3 RC4-MD5      1000 x      1   14.46s   1.56s
 SSLv3 RC4-MD5      1000 x      1   51.93s   1.62s 1024bit RSA
 SSLv3 RC4-SHA      1000 x      1   14.61s   1.83s
 SSLv3 DES-CBC-SHA  1000 x      1   14.70s   1.89s
-SSLv3 DES-CBC3-SHA 1000 x      1   15.16s   2.16s
 
 SSLv2 RC4-MD5      1000 x   1024   13.72s   1.27s
 SSLv3 NULL-MD5     1000 x   1024   14.79s   1.92s
 SSLv3 RC4-MD5      1000 x   1024   52.58s   2.29s 1024bit RSA
 SSLv3 RC4-SHA      1000 x   1024   15.39s   2.67s
 SSLv3 DES-CBC-SHA  1000 x   1024   16.45s   3.55s
-SSLv3 DES-CBC3-SHA 1000 x   1024   18.21s   5.38s
 
 SSLv2 RC4-MD5      1000 x  10240   18.97s   6.52s
 SSLv3 NULL-MD5     1000 x  10240   17.79s   5.11s
@@ -43,7 +41,6 @@ SSLv3 RC4-MD5      1000 x  10240   20.25s   7.90s
 SSLv3 RC4-MD5      1000 x  10240   58.26s   8.08s 1024bit RSA
 SSLv3 RC4-SHA      1000 x  10240   22.96s  11.44s
 SSLv3 DES-CBC-SHA  1000 x  10240   30.65s  18.41s
-SSLv3 DES-CBC3-SHA 1000 x  10240   47.04s  34.53s
 
 SSLv2 RC4-MD5      1000 x 102400   70.22s  57.74s
 SSLv3 NULL-MD5     1000 x 102400   43.73s  31.03s
@@ -51,7 +48,6 @@ SSLv3 RC4-MD5      1000 x 102400   71.32s  58.83s
 SSLv3 RC4-MD5      1000 x 102400  109.66s  59.20s 1024bit RSA
 SSLv3 RC4-SHA      1000 x 102400   95.88s  82.21s
 SSLv3 DES-CBC-SHA  1000 x 102400  173.22s 160.55s
-SSLv3 DES-CBC3-SHA 1000 x 102400  336.61s 323.82s
 
 What does this all mean?  Well for a server, with no session-id reuse, with
 a transfer size of 10240 bytes, using RC4-MD5 and a 512bit server key,
