diff --git a/release/src/router/openssl/doc/apps/ciphers.pod b/release/src/router/openssl/doc/apps/ciphers.pod
index 9540c9d..a486187 100644
--- a/release/src/router/openssl/doc/apps/ciphers.pod
+++ b/release/src/router/openssl/doc/apps/ciphers.pod
@@ -428,6 +428,7 @@ e.g. DES-CBC3-SHA. In these cases, RSA authentication is used.
 
 =head2 AES ciphersuites from RFC3268, extending TLS v1.0
 
+ TLS_RSA_WITH_AES_256_CBC_SHA            AES256-SHA
 
  TLS_DH_DSS_WITH_AES_128_CBC_SHA         DH-DSS-AES128-SHA
  TLS_DH_DSS_WITH_AES_256_CBC_SHA         DH-DSS-AES256-SHA
diff --git a/release/src/router/openssl/ssl/s3_lib.c b/release/src/router/openssl/ssl/s3_lib.c
index 0e9dfd9..1f7fcc8 100644
--- a/release/src/router/openssl/ssl/s3_lib.c
+++ b/release/src/router/openssl/ssl/s3_lib.c
@@ -1018,7 +1018,7 @@ OPENSSL_GLOBAL SSL_CIPHER ssl3_ciphers[] = {
      },
 
 /* Cipher 35 */
-    /*{
+    {
      1,
      TLS1_TXT_RSA_WITH_AES_256_SHA,
      TLS1_CK_RSA_WITH_AES_256_SHA,
@@ -1031,7 +1031,7 @@ OPENSSL_GLOBAL SSL_CIPHER ssl3_ciphers[] = {
      SSL_HANDSHAKE_MAC_DEFAULT | TLS1_PRF,
      256,
      256,
-     },*/
+     },
 /* Cipher 36 */
     {
      1,
diff --git a/release/src/router/openssl/ssl/t1_trce.c b/release/src/router/openssl/ssl/t1_trce.c
index c5ddea0..59b3289 100644
--- a/release/src/router/openssl/ssl/t1_trce.c
+++ b/release/src/router/openssl/ssl/t1_trce.c
@@ -191,7 +191,7 @@ static ssl_trace_tbl ssl_ciphers_tbl[] = {
     {0x0032, "TLS_DHE_DSS_WITH_AES_128_CBC_SHA"},
     {0x0033, "TLS_DHE_RSA_WITH_AES_128_CBC_SHA"},
     {0x0034, "TLS_DH_anon_WITH_AES_128_CBC_SHA"},
-    //{0x0035, "TLS_RSA_WITH_AES_256_CBC_SHA"},
+    {0x0035, "TLS_RSA_WITH_AES_256_CBC_SHA"},
     {0x0036, "TLS_DH_DSS_WITH_AES_256_CBC_SHA"},
     {0x0037, "TLS_DH_RSA_WITH_AES_256_CBC_SHA"},
     {0x0038, "TLS_DHE_DSS_WITH_AES_256_CBC_SHA"},
diff --git a/release/src/router/openssl/ssl/tls1.h b/release/src/router/openssl/ssl/tls1.h
index 5a5c402..82d0879 100644
--- a/release/src/router/openssl/ssl/tls1.h
+++ b/release/src/router/openssl/ssl/tls1.h
@@ -585,7 +585,7 @@ SSL_CTX_callback_ctrl(ssl,SSL_CTRL_SET_TLSEXT_TICKET_KEY_CB,(void (*)(void))cb)
 # define TLS1_TXT_DHE_RSA_WITH_AES_128_SHA               "DHE-RSA-AES128-SHA"
 # define TLS1_TXT_ADH_WITH_AES_128_SHA                   "ADH-AES128-SHA"
 
-//# define TLS1_TXT_RSA_WITH_AES_256_SHA                   "AES256-SHA"
+# define TLS1_TXT_RSA_WITH_AES_256_SHA                   "AES256-SHA"
 # define TLS1_TXT_DH_DSS_WITH_AES_256_SHA                "DH-DSS-AES256-SHA"
 # define TLS1_TXT_DH_RSA_WITH_AES_256_SHA                "DH-RSA-AES256-SHA"
 # define TLS1_TXT_DHE_DSS_WITH_AES_256_SHA               "DHE-DSS-AES256-SHA"
diff --git a/release/src/router/openssl/test/testfipsssl b/release/src/router/openssl/test/testfipsssl
index 876f09d..6dcfb32 100644
--- a/release/src/router/openssl/test/testfipsssl
+++ b/release/src/router/openssl/test/testfipsssl
@@ -11,7 +11,7 @@ else
   cert="$2"
 fi
 
-ciphers="DHE-RSA-AES256-SHA:DHE-DSS-AES256-SHA:EDH-RSA-DES-CBC3-SHA:EDH-DSS-DES-CBC3-SHA:DHE-RSA-AES128-SHA:DHE-DSS-AES128-SHA:EXP1024-DHE-DSS-DES-CBC-SHA:EXP1024-DES-CBC-SHA:EDH-RSA-DES-CBC-SHA:EDH-DSS-DES-CBC-SHA:DES-CBC-SHA:EXP-EDH-RSA-DES-CBC-SHA:EXP-EDH-DSS-DES-CBC-SHA:EXP-DES-CBC-SHA"
+ciphers="DHE-RSA-AES256-SHA:DHE-DSS-AES256-SHA:AES256-SHA:EDH-RSA-DES-CBC3-SHA:EDH-DSS-DES-CBC3-SHA:DHE-RSA-AES128-SHA:DHE-DSS-AES128-SHA:EXP1024-DHE-DSS-DES-CBC-SHA:EXP1024-DES-CBC-SHA:EDH-RSA-DES-CBC-SHA:EDH-DSS-DES-CBC-SHA:DES-CBC-SHA:EXP-EDH-RSA-DES-CBC-SHA:EXP-EDH-DSS-DES-CBC-SHA:EXP-DES-CBC-SHA"
 
 ssltest="../util/shlib_wrap.sh ./ssltest -F -key $key -cert $cert -c_key $key -c_cert $cert -cipher $ciphers"
 
