From: Robert Luberda <robert@debian.org>
Date: Tue, 23 Mar 2010 22:04:47 +0100
Subject: 00-Makefile

Makefile fixes.
---
 Makefile.in |   42 ++++++++++++++++++++++--------------------
 1 files changed, 22 insertions(+), 20 deletions(-)

diff --git a/Makefile.in b/Makefile.in
index d1ea853..60c10f9 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -69,9 +69,9 @@ NLS_DIR = $(PREFIX)/share/locale
 SYSCONFIG_DIR = @SYSCONFIG_DIR@
 
 # Compiler flags
-CFLAGS = @CFLAGS@ -Wall -Wstrict-prototypes -pipe -O2
+CFLAGS = @CFLAGS@ -Wall -Wstrict-prototypes -pipe
 DFLAGS = @DFLAGS@
-LFLAGS =  -s
+LFLAGS = @LDFLAGS@
 DFLAGS += -DSA_DIR=\"$(SA_DIR)\" -DSADC_PATH=\"$(SADC_PATH)\"
 DFLAGS += $(DFSENSORS)
 
@@ -239,8 +239,9 @@ else
 locales:
 endif
 
-nls/sysstat.pot: $(wildcard *.c)
-	$(XGETTEXT) -o $@ -k_ --msgid-bugs-address="sysstat <at> orange.fr" $^
+#.PHONY: nls/sysstat.pot
+nls/sysstat.pot: | $(wildcard *.c)
+	$(XGETTEXT) -o $@ -k_ --msgid-bugs-address="sysstat <at> orange.fr" $|
 
 # Phony targets
 .PHONY: clean distclean install install_base install_all uninstall \
@@ -273,9 +274,6 @@ ifeq ($(INSTALL_DOC),y)
 	$(INSTALL_DATA) $(MANGRPARG) man/nfsiostat.1 $(DESTDIR)$(MAN1_DIR)
 	rm -f $(DESTDIR)$(MAN1_DIR)/cifsiostat.1*
 	$(INSTALL_DATA) $(MANGRPARG) man/cifsiostat.1 $(DESTDIR)$(MAN1_DIR)
-ifeq ($(INSTALL_ISAG),y)
-	$(INSTALL_DATA) $(MANGRPARG) contrib/isag/isag.1 $(DESTDIR)$(MAN1_DIR)
-endif
 ifeq ($(COMPRESS_MANPG),y)
 	$(ZIP) $(DESTDIR)$(MAN8_DIR)/sa1.8
 	$(ZIP) $(DESTDIR)$(MAN8_DIR)/sa2.8
@@ -306,8 +304,15 @@ ifdef REQUIRE_NLS
 	done
 endif
 
-install_base: all sa1 sa2 sysstat.sysconfig install_man install_nls \
-	contrib/isag/isag
+install_isag: contrib/isag/isag
+ifeq ($(INSTALL_ISAG),y)
+	mkdir -p $(DESTDIR)$(BIN_DIR)
+	mkdir -p $(DESTDIR)$(MAN1_DIR)
+	$(INSTALL_BIN) contrib/isag/isag $(DESTDIR)$(BIN_DIR)
+	$(INSTALL_DATA) $(MANGRPARG) contrib/isag/isag.1 $(DESTDIR)$(MAN1_DIR)
+endif
+
+install_base: all sa1 sa2 sysstat.sysconfig install_man install_nls install_isag
 	mkdir -p $(DESTDIR)$(SA_LIB_DIR)
 	mkdir -p $(DESTDIR)$(SA_DIR)
 ifeq ($(CLEAN_SA_DIR),y)
@@ -316,7 +321,7 @@ ifeq ($(CLEAN_SA_DIR),y)
 	-rmdir --ignore-fail-on-non-empty $(DESTDIR)$(SA_DIR)/[0-9]?????
 endif
 	mkdir -p $(DESTDIR)$(BIN_DIR)
-	mkdir -p $(DESTDIR)$(DOC_DIR)
+	#mkdir -p $(DESTDIR)$(DOC_DIR)
 	mkdir -p $(DESTDIR)$(SYSCONFIG_DIR)
 	$(INSTALL_BIN) sa1 $(DESTDIR)$(SA_LIB_DIR)
 	$(INSTALL_BIN) sa2 $(DESTDIR)$(SA_LIB_DIR)
@@ -328,18 +333,15 @@ endif
 	$(INSTALL_BIN) pidstat $(DESTDIR)$(BIN_DIR)
 	$(INSTALL_BIN) nfsiostat $(DESTDIR)$(BIN_DIR)
 	$(INSTALL_BIN) cifsiostat $(DESTDIR)$(BIN_DIR)
-ifeq ($(INSTALL_ISAG),y)
-	$(INSTALL_BIN) contrib/isag/isag $(DESTDIR)$(BIN_DIR)
-endif
 	$(INSTALL_DATA) sysstat.ioconf $(DESTDIR)$(SYSCONFIG_DIR)
 	$(INSTALL_DATA) sysstat.sysconfig $(DESTDIR)$(SYSCONFIG_DIR)/sysstat
 ifeq ($(INSTALL_DOC),y)
-	$(INSTALL_DATA) CHANGES $(DESTDIR)$(DOC_DIR)
-	$(INSTALL_DATA) COPYING $(DESTDIR)$(DOC_DIR)
-	$(INSTALL_DATA) CREDITS $(DESTDIR)$(DOC_DIR)
-	$(INSTALL_DATA) README $(DESTDIR)$(DOC_DIR)
-	$(INSTALL_DATA) FAQ $(DESTDIR)$(DOC_DIR)
-	$(INSTALL_DATA) *.lsm $(DESTDIR)$(DOC_DIR)
+#	$(INSTALL_DATA) CHANGES $(DESTDIR)$(DOC_DIR)
+#	$(INSTALL_DATA) COPYING $(DESTDIR)$(DOC_DIR)
+#	$(INSTALL_DATA) CREDITS $(DESTDIR)$(DOC_DIR)
+#	$(INSTALL_DATA) README $(DESTDIR)$(DOC_DIR)
+#	$(INSTALL_DATA) FAQ $(DESTDIR)$(DOC_DIR)
+#	$(INSTALL_DATA) *.lsm $(DESTDIR)$(DOC_DIR)
 endif
 
 install_all: install_base cron/crontab sysstat \
@@ -479,7 +481,7 @@ clean:
 	rm -f sadc sar sadf iostat mpstat pidstat nfsiostat cifsiostat *.o *.a core TAGS
 	find nls -name "*.gmo" -exec rm -f {} \;
 
-almost-distclean: clean nls/sysstat.pot
+almost-distclean: clean
 	rm -f sa1 sa2 sysstat cron/crontab version.h sysconfig.h
 	rm -f sysstat.sysconfig cron/sysstat.crond cron/sysstat.cron.daily
 	rm -f cron/sysstat.cron.hourly cron/sysstat.crond.sample cron/sysstat.crond.sample.in
-- 
