diff -ur ./mtd-utils-20090606/recv_image.c ./mtd-utils-20090606-new/recv_image.c
--- ./mtd-utils-20090606/recv_image.c	2009-06-05 12:41:47.000000000 -0400
+++ ./mtd-utils-20090606-new/recv_image.c	2015-05-19 11:31:34.063052615 -0400
@@ -5,6 +5,7 @@
 #include <error.h>
 #include <stdio.h>
 #define __USE_GNU
+#define __USE_MISC
 #include <netdb.h>
 #include <stdlib.h>
 #include <string.h>
diff -ur ./mtd-utils-20090606.orig/Makefile ./mtd-utils-20090606/Makefile
--- ./mtd-utils-20090606.orig/Makefile	2009-06-05 09:41:47.000000000 -0700
+++ ./mtd-utils-20090606/Makefile	2015-08-20 03:29:53.402360707 -0700
@@ -15,7 +15,6 @@
 	jffs2dump \
 	nftldump nftl_format docfdisk \
 	rfddump rfdformat \
-	serve_image recv_image \
 	sumtool #jffs2reader
 
 SYMLINKS =

diff -purN ./mtd-utils-20090606.orig/mkfs.jffs2.c ./mtd-utils-20090606/mkfs.jffs2.c
--- ./mtd-utils-20090606.orig/mkfs.jffs2.c	2009-06-06 00:41:47.000000000 +0800
+++ ./mtd-utils-20090606/mkfs.jffs2.c	2020-06-16 19:14:01.413829608 +0800
@@ -74,6 +74,7 @@
 #undef crc32
 #include "crc32.h"
 #include "rbtree.h"
+#include <sys/sysmacros.h>
 
 /* Do not use the weird XPG version of basename */
 #undef basename
diff -purN ./mtd-utils-20090606.orig/mkfs.ubifs/devtable.c ./mtd-utils-20090606/mkfs.ubifs/devtable.c
--- ./mtd-utils-20090606.orig/mkfs.ubifs/devtable.c	2009-06-06 00:41:47.000000000 +0800
+++ ./mtd-utils-20090606/mkfs.ubifs/devtable.c	2020-06-16 19:23:16.967082829 +0800
@@ -44,6 +44,7 @@
  * for more information about what the device table is.
  */
 
+#include <sys/sysmacros.h>
 #include "mkfs.ubifs.h"
 #include "hashtable/hashtable.h"
 #include "hashtable/hashtable_itr.h"
diff -ur ./mtd-utils-20090606.org/mkfs.ubifs/hashtable/hashtable_itr.c ./mtd-utils-20090606/mkfs.ubifs/hashtable/hashtable_itr.c
--- ./mtd-utils-20090606.org/mkfs.ubifs/hashtable/hashtable_itr.c
+++ ./mtd-utils-20090606/mkfs.ubifs/hashtable/hashtable_itr.c
@@ -33,18 +33,6 @@
     }
     return itr;
 }
-
-/*****************************************************************************/
-/* key      - return the key of the (key,value) pair at the current position */
-/* value    - return the value of the (key,value) pair at the current position */
-
-void *
-hashtable_iterator_key(struct hashtable_itr *i)
-{ return i->e->k; }
-
-void *
-hashtable_iterator_value(struct hashtable_itr *i)
-{ return i->e->v; }
 
 /*****************************************************************************/
 /* advance - advance the iterator to the next element
diff -ur ./mtd-utils-20090606_orig/mkfs.ubifs/hashtable/hashtable_itr.h ./mtd-utils-20090606/mkfs.ubifs/hashtable/hashtable_itr.h
--- ./mtd-utils-20090606_orig/mkfs.ubifs/hashtable/hashtable_itr.h
+++ ./mtd-utils-20090606/mkfs.ubifs/hashtable/hashtable_itr.h
@@ -28,7 +28,7 @@
 /* hashtable_iterator_key
  * - return the value of the (key,value) pair at the current position */
 
-extern inline void *
+static inline void *
 hashtable_iterator_key(struct hashtable_itr *i)
 {
     return i->e->k;
@@ -37,7 +37,7 @@
 /*****************************************************************************/
 /* value - return the value of the (key,value) pair at the current position */
 
-extern inline void *
+static inline void *
 hashtable_iterator_value(struct hashtable_itr *i)
 {
     return i->e->v;
diff -purN ./mtd-utils-20090606.orig/mkfs.ubifs/mkfs.ubifs.c ./mtd-utils-20090606/mkfs.ubifs/mkfs.ubifs.c
--- ./mtd-utils-20090606.orig/mkfs.ubifs/mkfs.ubifs.c	2009-06-06 00:41:47.000000000 +0800
+++ ./mtd-utils-20090606/mkfs.ubifs/mkfs.ubifs.c	2020-06-16 19:31:18.265064661 +0800
@@ -20,6 +20,9 @@
  *          Zoltan Sogor
  */
 
+#define _GNU_SOURCE
+#include <sys/sysmacros.h>
+#include <sys/types.h>
 #include "mkfs.ubifs.h"
 
 #define PROGRAM_VERSION "1.3"
diff -purN ./mtd-utils-20090606.orig/ubi-utils/old-utils/src/libubi.c ./mtd-utils-20090606/ubi-utils/old-utils/src/libubi.c
--- ./mtd-utils-20090606.orig/ubi-utils/old-utils/src/libubi.c	2009-06-06 00:41:47.000000000 +0800
+++ ./mtd-utils-20090606/ubi-utils/old-utils/src/libubi.c	2020-06-16 19:21:53.065911676 +0800
@@ -30,6 +30,7 @@
 #include <errno.h>
 #include <unistd.h>
 #include <sys/ioctl.h>
+#include <sys/sysmacros.h>
 #include <limits.h>
 #include "libubi.h"
 #include "libubi_int.h"
diff -purN ./mtd-utils-20090606.orig/ubi-utils/src/libmtd.c ./mtd-utils-20090606/ubi-utils/src/libmtd.c
--- ./mtd-utils-20090606.orig/ubi-utils/src/libmtd.c	2009-06-06 00:41:47.000000000 +0800
+++ ./mtd-utils-20090606/ubi-utils/src/libmtd.c	2020-06-16 19:19:25.226990941 +0800
@@ -32,6 +32,7 @@
 #include <sys/stat.h>
 #include <sys/ioctl.h>
 #include <mtd/mtd-user.h>
+#include <sys/sysmacros.h>
 
 #include <libmtd.h>
 #include "libmtd_int.h"
diff -purN ./mtd-utils-20090606.orig/ubi-utils/src/libmtd_legacy.c ./mtd-utils-20090606/ubi-utils/src/libmtd_legacy.c
--- ./mtd-utils-20090606.orig/ubi-utils/src/libmtd_legacy.c	2009-06-06 00:41:47.000000000 +0800
+++ ./mtd-utils-20090606/ubi-utils/src/libmtd_legacy.c	2020-06-16 19:20:41.767766233 +0800
@@ -32,6 +32,7 @@
 #include <sys/stat.h>
 #include <sys/ioctl.h>
 #include <mtd/mtd-user.h>
+#include <sys/sysmacros.h>
 
 #include <libmtd.h>
 #include "libmtd_int.h"
diff -purN ./mtd-utils-20090606.orig/ubi-utils/src/libubi.c ./mtd-utils-20090606/ubi-utils/src/libubi.c
--- ./mtd-utils-20090606.orig/ubi-utils/src/libubi.c	2009-06-06 00:41:47.000000000 +0800
+++ ./mtd-utils-20090606/ubi-utils/src/libubi.c	2020-06-16 19:18:14.259644442 +0800
@@ -33,6 +33,7 @@
 #include <libubi.h>
 #include "libubi_int.h"
 #include "common.h"
+#include <sys/sysmacros.h>
 
 #define PROGRAM_NAME "libubi"
 
